### PR TITLE
Use `eth_call` For Fetching Current Block Header Information

### PR DESCRIPTION
**Pull request by nlordell**
This PR is an attempt to reduce the number of "No state available for block 0x..." errors that we see.

Specifically, in looking at the [Nethermind code](https://github.com/NethermindEth/nethermind/blob/6d0409c2c98c3a311c29fd90491666f33e7d5f0d/src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthModule.TransactionExecutor.cs#L51-L62), it appears that this happens when a block header is available, but not its state. Interestingly, when querying for the latest block, [it only searches for the block header](https://github.com/NethermindEth/nethermind/blob/6d0409c2c98c3a311c29fd90491666f33e7d5f0d/src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs#L396-L400).

To work around this, instead of querying block header information the traditional way, we use an `eth_call` with carefully crafted bytecode to read this state within an execution context (see `FetchBlock.sol`). This ensures:
- That the block state that it is returning is available (since the call context that is running has access to it).
- It is atomic (so no race conditions of one node handling the reading of the block number and another handling the reading of the hash).

This also reduces the amount of data that we are polling from the node (full block headers are actually quite large), so it may be a reasonable change in and of itself.

As a side benefit, we reduce a bunch of weird complexity around the `CurrentBlockStream` returning a block without a number. So 🎉.

### Test Plan

Run the blockstream locally and check on Etherscan.io that the latest block being reported is correct:
```
% cargo test -p shared -- current_block::tests::mainnet --ignored --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.26s
     Running unittests src/lib.rs (target/debug/deps/shared-b80690e295f197ca)

running 1 test
new block number 15998074
new block number 15998075
new block number 15998076
test current_block::tests::mainnet ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 389 filtered out; finished in 26.43s
```

image

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]
Arguably the correct fix would be for nethermind to have a way to get the "current block that has state" but this is a good workaround. It does hinge on "pending" to work right so we should probably test this with the different nodes we use.

**Comment by nlordell** [comment]
> It does hinge on "pending" to work right so we should probably test this with the different nodes we use.

I tried it with Geth and Nethermind and both worked.

File: crates/contracts/src/lib.rs [hunk 1]
```diff
@@ -60,6 +60,8 @@ pub mod support {
         AnyoneAuthenticator;
         PhonyERC20;
         Trader;
+
```
Comment (bh2smith) [inline crates/contracts/src/lib.rs:63]:
```suggestion
```

File: crates/shared/src/current_block.rs [hunk 2]
```diff
@@ -54,17 +57,12 @@ impl<T: Ord> RangeInclusive<T> {
 pub async fn current_block_stream(
```
Comment (bh2smith) [inline crates/shared/src/current_block.rs:57]: Really nice simplification inside this method!

File: crates/solver/src/driver.rs [hunk 3]
```diff
@@ -30,11 +30,8 @@ use model::{
 use num::{rational::Ratio, BigInt, BigRational, ToPrimitive};
 use primitive_types::{H160, U256};
 use shared::{
-    current_block::{self, CurrentBlockStream},
-    ethrpc::Web3,
-    http_solver::model::SolverRunError,
-    recent_block_cache::Block,
-    tenderly_api::TenderlyApi,
+    current_block::CurrentBlockStream, ethrpc::Web3, http_solver::model::SolverRunError,
+    recent_block_cache::Block, tenderly_api::TenderlyApi,
```
Comment (bh2smith) [inline crates/solver/src/driver.rs:34]: I guess cargo format didn't prefer the prettier layout? Kinda like the way it was before.

File: crates/shared/src/macros.rs [hunk 4]
```diff
@@ -40,6 +40,13 @@ macro_rules! dummy_contract {
     };
 }
 
+#[macro_export]
```
Comment (bh2smith) [inline crates/shared/src/macros.rs:43]: I'm not able to tell where this is explicitly used...

**Review by bh2smith — APPROVED** [review]
Really really cook stuff here and a lot of nice implications overall.

File: crates/solver/src/driver.rs (hunk 3 above)
Comment (nlordell) [reply crates/solver/src/driver.rs:34]: We are all at the mercy of auto-formatting... 😛

File: crates/shared/src/macros.rs (hunk 4 above)
Comment (nlordell) [reply crates/shared/src/macros.rs:43]: Its used in `current_block.rs` (a bit discretely - the `eth_call` that we do there needs the bytecode of the `FetchBlock.sol` contract).

File: crates/shared/src/current_block.rs [hunk 5]
```diff
@@ -110,42 +95,58 @@ pub async fn current_block_stream(
...
 impl BlockRetrieving for Web3 {
-    async fn current_block(&self) -> Result<Block> {
-        self.eth()
-            .block(BlockId::Number(BlockNumber::Latest))
+    async fn current_block(&self) -> Result<BlockInfo> {
+        let return_data = self
+            .eth()
+            .call(
+                CallRequest {
+                    data: Some(bytecode!(FetchBlock)),
```
Comment (nlordell) [inline crates/shared/src/current_block.rs:126]: @bh2smith here is where it gets used.

File: crates/shared/src/maintenance.rs [hunk 6]
```diff
@@ -43,7 +43,7 @@ impl Maintaining for ServiceMaintenance {
 }
 
 impl ServiceMaintenance {
-    pub async fn run_maintenance_for_blocks(self, blocks: impl Stream<Item = Block>) {
+    async fn run_maintenance_for_blocks(self, blocks: impl Stream<Item = BlockInfo>) {
```
Comment (nlordell) [inline crates/shared/src/maintenance.rs:46]: Should never have been `pub` in the first place... Oops!

**Review by MartinquaXD — APPROVED** [review]
Nice cleanup! 🧹

**Comment by nlordell** [comment]
> Arguably the correct fix would be for nethermind to have a way to get the "current block that has state" but this is a good workaround

@vkgnosis - I made a follow-up PR where I make using this strategy a configurable flag instead of using it always.

**Pull request merged**